model/implementation/DeliveryExecutionStateService.php

```diff
@@ -88,6 +92,32 @@ public function waitExecution(DeliveryExecution $deliveryExecution)
         $result = false;
 
         if (self::STATE_TERMINATED != $executionState && self::STATE_COMPLETED != $executionState) {
+
+            if ($session = $this->getTestSessionService()->getTestSession($deliveryExecution)) {
+
+                $wasPausedAt = null;
+
+                $resultServer = \taoResultServer_models_classes_ResultServerStateFull::singleton();
+                $vars = array_reverse($resultServer->getVariables($session->getSessionId()));
+                foreach ($vars as $arr) {
+                    $var = reset($arr)->variable;
+                    if (substr($var->identifier, 0, strlen('TEST_PAUSE')) == 'TEST_PAUSE') {
+                        $trace = $trace = json_decode($var->trace, true);
+                        $wasPausedAt = (new DateTimeImmutable())->setTimestamp(\tao_helpers_Date::getTimeStamp($trace['timestamp']));
+                        break;
+                    }
+                }
+
+                if ($wasPausedAt && $this->hasOption('termination_delay_after_pause')) {
+                    $delay = $this->getOption('termination_delay_after_pause');
+                    if ($wasPausedAt->add(new DateInterval($delay)) < (new DateTime())) {
+                        $this->terminateExecution($deliveryExecution, null, ['test' => TestSessionProcessor::TEST_CODE_INCOMPLETE]);
+
+                        return false;
+                    }
+                }
+            }
+
```

Comment:
Could you move all this stuff from `waitExecution()` function to separate protected method (something like `isExpired()`)?